### DISCUSSION TITLE
[Issue]HW Usage during lunching of the simulator

### DISCUSSION THREAD

**Discussion by AlaaElnagar** [langroodi/Adaptive-AUTOSAR#30] [Q&A]
Hello, 

Could you please inform me if it applicable to run the simulator without the usage of the HW 
![Screenshot 2023-11-25 080011](https://github.com/langroodi/Adaptive-AUTOSAR/assets/35043356/a541227f-801a-4818-899c-d92da164bc43)

**Comment by langroodi** [comment]
Hi!

No, it is not required to have the OBD scanner to run the simulation. But just be aware that in that case the DM will fail. To avoid that, you have two options:

1. Create a virtual serial port at `/dev/ttyUSB0`;
2. Remove `mSerialCommunication` from [DM](https://github.com/langroodi/Adaptive-AUTOSAR/blob/master/src/application/platform/diagnostic_manager.cpp).

In my [demo video](https://youtu.be/OFYA-raVrWE?t=3379), I also explained that.

**Comment by GokulKumar1103** [comment]
Dear Armin,
I am currently working on resolving the "**Starting OBD-II emulator failed**" issue. To better understand where the failure occurs, I have set breakpoints in the relevant files. Upon investigation, I identified that within the **TryStart( )** method, we are invoking the **trySetupCommunication( )** function, which is returning false.

Inside the **trySetupCommunication( )** function, the process attempts to open the **serial** port (/dev/ttyUSB0), but the command returns -1.
I have tried the first method you suggested, which involves **creating a virtual serial port at /dev/ttyUSB0**, but  the issue was not solved .
I have attached the output for your reference. Could you kindly review it and let me know where the issue might be originating or suggest any further steps I should take to resolve this?
![Serial_Port](https://github.com/user-attachments/assets/abf53058-5467-444f-bf4b-e26340a58050)

**Comment by GokulKumar1103** [comment]
**Executed the first method of hardware independence: Creating a virtual serial port at /dev/ttyUSB0** 
In the previous stage, I encountered an issue regarding the usage of the virtually created serial port. Upon checking, I discovered that another process was using that port, which prevented the adaptive application from opening the communication successfully. After killing that process and recreating the virtual port, I received the desired output. At this point, the issue "Starting OBD-II emulator failed" was no longer printed, and in the **"trySetupCommunication()"** function, the serial port was successfully opened.

Now, could you please help me understand how to access the other APIs of the VOLVO vehicle? What is the next step in learning about the Adaptive AUTOSAR Linux Simulator?
![Screenshot 2025-04-07 205440](https://github.com/user-attachments/assets/5aaf16c7-a89a-4663-ab52-97914d4e5ea5)

**Reply by langroodi** [reply]
> Now, could you please help me understand how to access the other APIs of the VOLVO vehicle?

Please follow [this tutorial](https://developer.volvocars.com/apis/docs/getting-started/#prerequisites) from Volvo in order to make a token to interact with the Extended Vehicle API. However, Volvo moved to API v2 and I am not sure whether they still support the v1 (which is used within this project) or not.

> What is the next step in learning about the Adaptive AUTOSAR Linux Simulator?

Can you please make it more specific? What is your goal of learning the Adaptive AUTOSAR? Then I may be able to provide you more specific answer.

**Comment by GokulKumar1103** [comment]
Thank you for your response.
I will go through the link you shared regarding the Extended Vehicle API V2 and will keep you updated accordingly.

I am currently exploring Adaptive AUTOSAR, and as I am new to it, my primary goal is to gain a solid understanding of Adaptive AUTOSAR applications, C++ development, and working within a Linux environment.

Today, I encountered an issue regarding how to send diagnostic commands via a virtual serial port and receive responses. Like , if I want to retrieve engine status or speed, how can this be achieved through the serial port?

Please note that I am not using any hardware (such as an OBD-II interface). I would appreciate it if you could provide some insights or guidance on how to proceed.

**Comment by langroodi** [comment]
> I am currently exploring Adaptive AUTOSAR, and as I am new to it, my primary goal is to gain a solid understanding of Adaptive AUTOSAR applications, C++ development, and working within a Linux environment.

Your primary goal is a bit broad for me. Focusing only on the ARA part, I would recommend to start with the resources mentioned in [this part of the video](https://youtu.be/OFYA-raVrWE?feature=shared&t=121) to get the big picture of the platform. Then the next step would be to pick one specific ARA domain (a list is mentioned in [the project wiki](https://github.com/langroodi/Adaptive-AUTOSAR/wiki)) to go deeper into that. In reality, ARA developers do not focus on several domains because it is too much. Some work on `ara::com`, the other work on `ara::iam`, etc.

For each domain, there is a Requirement Specification (RS) document that explains the functional requirement for that domain and there is SoftWare Specification (SWS) document that explains the software design. These two are a good starting point to explore any new domain within the platform.

> Today, I encountered an issue regarding how to send diagnostic commands via a virtual serial port and receive responses. Like , if I want to retrieve engine status or speed, how can this be achieved through the serial port?

Let's assume that you want to get the engine speed via OBD-II. Each command in OBD-II has a PID (please refer to [this link](https://en.wikipedia.org/wiki/OBD-II_PIDs)) and in this case the PID is `0x0C`. So, if you send this PID (which is one byte) to that virtual serial point, would the simulator returns back the speed in RPM? The answer is No! To find out why, please refer to the [OBD-II emulator architecture](https://github.com/langroodi/OBD-II-Emulator).

Referring back to the first question I've asked before: _what is your learning goal here?_

- Do you want to learn about OBD-II (which is not related to ARA)? Then you need to change the OBD-II emulator;
- Do you want to learn about CAN (which is also not related to ARA)? Then you need to change the OBD-II emulator again;
- Do you want to only learn about ARA? Then you can skip the serial port part totally and change the ARA simulator in order to receive the PID from a different way (e.g., an IPC mechanism).

**Reply by GokulKumar1103** [reply]
Thank you very much for your detailed explanation and for taking the time to guide me.

Based on your suggestions, I will refocus my efforts on the ARA aspect and choose a specific domain to explore in depth. As advised, I will begin referring to the Requirement Specification (RS) and Software Specification (SWS) documents for that particular domain to build a stronger foundation.
I would like to know if it is possible to retrieve engine speed without relying on a real OBD-II interface.
I believe my primary learning objective is to gain a deeper understanding of ARA.

**Comment by GokulKumar1103** [comment]
Hi Armin,

I’ve been analyzing the Diagnostics (Diag 0x8001) and Vehicle Identification (0x01) payload types in the DoIP implementation and wanted to share my understanding along with a few questions.

In both doip_client and doip_server files, I found useful information on how requests are structured and sent, and how responses are handled. Based on my understanding, the client sends requests using the send() function with the relevant payload structure. For example:

Vehicle Identification Payload
{0x02, 0xFD, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00}

Diagnostic Message Payload
{0x02, 0xFD, 0x80, 0x01, 0x00, 0x00, 0x00, 0x07, 0x00, 0x00, 0x00, 0x00, 0x22, 0xF5, 0x0D}
(Here, 0x22 is the SID for ReadDataByIdentifier and 0xF50D is the DID for Average Speed)

The process flow I've observed is:

client send request  -> server receive request & process response  -> server send response  -> client receive response

After sending the request using tcp_send(), the server receives it via tcp_receive() and processes it. I’m attaching a snippet related to payload type 0x01 for reference
![image](https://github.com/user-attachments/assets/98e5b864-bea7-4cd4-a1ea-65e7e316ab57)

The response is then sent back to the client, where functions like handleVehicleIdResponse() and handleDiagMessageAck() process it. However:

--> I couldn’t trace what happens after these handle functions execute. Where does the flow go after processing the response?

--> I’ve been creating custom request payloads manually and rebuilding the software for each change. Is there a way to avoid hardcoding each request ?

-->Once "Extended Vehicle AA is discovered", I receive a message saying "Alive supervision is expired on MachineFG" after about a minute, which leads to process termination. How can this timeout be prevented or extended?

Looking forward to your insights.

**Reply by langroodi** [reply]
Good to see that you are still progressing :)

> I couldn’t trace what happens after these handle functions execute. Where does the flow go after processing the response?

It is kinda understandable why!

- `handleVehicleIdResponse` only sets the DoIP *Logical Address*  which is required for sending diagnostic messages via DoIP.

- On the other hand, `handleDiagMessageAck` triggers a callback. This callback ends up in `TryGetResponseAsync` method within `ObdToDoipConverter`.

How? You need to check the source code of `ObdEmulator::ObdService`!
Why? Because it is an async communication in which the Inversion of Control a.k.a. IoC pattern is used to make it possible!

> I’ve been creating custom request payloads manually and rebuilding the software for each change. Is there a way to avoid hardcoding each request ?

In the place where you made the payload hard-coded, you can open a network socket to get the payload from TCP/IP port, or you can read the payload from a file there, or you can get it from the console via `std::cin`. I go for the second option to read it from a file. It is easy and you don't need to build the project each time.

> Once "Extended Vehicle AA is discovered", I receive a message saying "Alive supervision is expired on MachineFG" after about a minute, which leads to process termination. How can this timeout be prevented or extended?

The message is coming from the Platform Health Management (PHM). As the name indicates, it is responsible for monitoring the health of an Adaptive Application rather than terminating that. When it detects that an application is not healthy, it triggers a `RecoveryAction` and in this case the `LogRecoveryAction` only logs the monitoring event and does not terminate the application.

So, if the application is terminated, that is nothing to do with the PHM and disabling the monitoring only hides the problem rather than solving that.
While if the PHM is reporting a false alarm, then you can change the configuration in the `health_monitoring_manifest.arxml` file, but I recommend to read first about the *Alive supervision* to get a better understanding about the parameters you want to change there.

Have fun with your exploration ;)

**Comment by GokulKumar1103** [comment]
Thanks, Armin, for your detailed explanation. 
I now understand the purpose of **handleVehicleIdResponse** and **handleDiagMessageAck**.
After executing the handler function, control flows to the **obd_emulator.cpp** file in the OBD project, where the **sendResponseAsync()** function is called. Inside that, the **TrySendAsync( )** method sends the payload data out through the serial port.

Based on my understanding, I have the following points:

 Questions:
1) From where does the DoIP request or OBD PID originate?
--> As per my understanding, the client initiates the request, and the server processes it by querying the VOLVO API using curl, and then responds back to the client. Is that correct?

2) Should we fetch all PID values  from the VOLVO API at once, or only for each request individually?
--> In your demo using OBD Doctor, when you clicked "Send," all relevant values appeared. Does this mean all PIDs were fetched together?

3) I plan to start with SOME/IP. Which communication method should I choose?
--> The project supports **PUB/SUB**, **RPC**, and **(SD)**. Since I’m just getting started, could you recommend which one is best suited for a beginner?  
Thanks again for your support and guidance.

**Reply by langroodi** [reply]
No worries!

> As per my understanding, the client initiates the request, and the server processes it by querying the VOLVO API using curl, and then responds back to the client. Is that correct?

Partially yes! As far as I can remember, only for 5 PIDs, the request is forwarded towards the API via Curl. For the rest, the request is being answered by some dummy values inside of the adaptive application itself. The reason for that was the API v1 was also not supporting all the PIDs.

> In your demo using OBD Doctor, when you clicked "Send," all relevant values appeared. Does this mean all PIDs were fetched together?

No, they are processed separately.

> The project supports PUB/SUB, RPC, and (SD). Since I’m just getting started, could you recommend which one is best suited for a beginner?

I wouldn't categorize that way. I recommend to start learning the idea behind all of these three pattern (what are they and why they exist). Probably the RPC would be the most familiar one. Then for the next step, you can dive into the implementation details of each pattern (which each of them has its own implementation challenge).

**Reply by GokulKumar1103** [reply]
Thanks , Armin for your explanation .

> Partially yes! As far as I can remember, only for 5 PIDs, the request is forwarded towards the API via Curl. For the rest, the request is being answered by some dummy values inside of the adaptive application itself. The reason for that was the API v1 was also not supporting all the PIDs.

> > ### **Yes I saw there was curl request was only for 5 PIDs and rest of them was some defaulat response value was feed ,so my assumption was clarified here.**

> No, they are processed separately.

> > ### **Oh, Okay . But when I saw the DEMO it seems to me like all the OBD requests are sent at the same time and their respective Response was came to serial port -> CAN  -> OBD** .

> I wouldn't categorize that way. I recommend to start learning the idea behind all of these three pattern (what are they and why they exist). Probably the RPC would be the most familiar one. Then for the next step, you can dive into the implementation details of each pattern (which each of them has its own implementation challenge).

> ### **I studied the SOME/IP Specification document and got some understanding . SOME/IP SD works on SOME/IP with UDP . You mentioned in a reply that , changing the IP Address from 127.0.0.1 (NIC Loopback IP ) to different one means we can saw the output . If that correct where I need to change like DM , Extended Vehicle .In this project we are not using the PUB/SUB this thing you already mentioned  . 
I checked the datahandling through tcpdumb on port 8080 and 8081 ,where the Socket RPC client send the 3 packets of data to Socket RPC server which seems to contain the header as first packet , StateManagement second packet and machineFG startup as third . I couldn't know what it says** .
